File: src/main/java/com/example/RESTAPIarticle/service/ArticleServiceImpl.java
package com.example.RESTAPIarticle.service;

import com.example.RESTAPIarticle.dao.ArticleDAO;
import com.example.RESTAPIarticle.dao.ContentDAO;
import com.example.RESTAPIarticle.entity.Article;
import com.example.RESTAPIarticle.entity.ArticleContent;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
@RequiredArgsConstructor
public class ArticleServiceImpl implements ArticleService{

    private final ArticleDAO articleDAO;
    private final ContentDAO contentDAO;
    private final Comparator<? super Article> ArticleComparator;

    @Override
    public List<Article> findAllOrderByDateDesc() {
        return articleDAO.findAll(Sort.by(Sort.Direction.DESC, "date"));
    }

    @Override
    public Article findById(int theId) {
        return articleDAO.findById(theId).orElseThrow();
    }

    @Override
    public List<Article> findAllByKeyword(String theKeyword) {
        List<Article> articles = new ArrayList<>();
        for(ArticleContent tempContent : contentDAO.findAllByTextContainsOrTitleContains(theKeyword,theKeyword)){
            articles.addAll(articleDAO.findAllByContentOrderByDateDesc(tempContent));
        }
        articles.sort(Collections.reverseOrder(ArticleComparator));
        return articles;
    }

    @Override
    public Article save(Article theArticle) {
        return articleDAO.save(theArticle);
    }

    @Override
    public void deleteById(int theId) {
        articleDAO.deleteById(theId);
    }
}

File: src/main/java/com/example/RESTAPIarticle/service/ArticleServiceImpl.kt
package com.example.RESTAPIarticle.service

import com.example.RESTAPIarticle.entity.Article
import com.example.RESTAPIarticle.errors.ArticleNotFoundException
import com.example.RESTAPIarticle.repository.ArticleRepository
import com.example.RESTAPIarticle.repository.ContentRepository
import lombok.RequiredArgsConstructor
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
@RequiredArgsConstructor
class ArticleServiceImpl(
    private val articleRepository: ArticleRepository,
    private val contentRepository: ContentRepository
) : ArticleService {

    override fun findAllOrderByDateDesc(): List<Article> {
        return articleRepository.findAll(Sort.by(Sort.Direction.DESC, "date"))
    }

    override fun findById(theId: Int): Article {
        return articleRepository.findById(theId)
            .orElseThrow { ArticleNotFoundException("Article id not found - $theId") }
    }

    override fun findAllByKeyword(theKeyword: String): List<Article> {
        val articles: List<Article> = articleRepository.findAll(Sort.by(Sort.Direction.DESC, "date"))
        return articles.stream()
            .filter { article: Article ->
                contentRepository.findAllByTextContainsOrTitleContains(
                    theKeyword,
                    theKeyword
                )!!.contains(article.content)
            }
            .collect(Collectors.toList())
    }

    override fun save(theArticle: Article): Article {
        return articleRepository.save(theArticle)
    }

    override fun deleteById(theId: Int) {
        articleRepository.deleteById(theId)
    }
}
